src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -575,6 +613,121 @@ fn max_slice_length<'p, 'a: 'p, 'tcx: 'a, I>(
+                            Some(IntRange { range: lo..=(hi - offset), ty })
+                        };
+                    }
+                }
+                None
+            }
+            ConstantValue(val) => {
+                let ty = val.ty;
+                if let Some(val) = val.assert_bits(tcx, ty::ParamEnv::empty().and(ty)) {
+                    let bias = IntRange::signed_bias(tcx, ty);
+                    let val = val ^ bias;
+                    Some(IntRange { range: val..=val, ty })
+                } else {
+                    None
+                }
+            }
+            Single | Variant(_) | Slice(_) => {
+                None
+            }
+        }
+    }
+
+    fn signed_bias(tcx: TyCtxt<'_, 'tcx, 'tcx>, ty: Ty<'tcx>) -> u128 {
+        match ty.sty {
+            ty::TyInt(ity) => {
+                let bits = Integer::from_attr(tcx, SignedInt(ity)).size().bits() as u128;
+                1u128 << (bits - 1)
+            }
+            _ => 0
+        }
+    }
+
+    fn into_inner(self) -> (u128, u128) {
+        self.range.into_inner()
+    }
+}
+
+/// Given a pattern in a `match` and a collection of ranges corresponding to the
+/// domain of values of a type (say, an integer), return a new collection of ranges
+/// corresponding to those ranges minus the ranges covered by the pattern.
+fn ranges_subtract_pattern<'a, 'tcx>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
+                                     pat_ctor: &Constructor<'tcx>,
+                                     ranges: Vec<Constructor<'tcx>>)
+                                     -> Vec<Constructor<'tcx>> {
+    if let Some(pat_interval) = IntRange::from_ctor(cx.tcx, pat_ctor) {
+        let mut remaining_ranges = vec![];
+        let mut ranges: Vec<_> = ranges.into_iter().filter_map(|r| {
+            IntRange::from_ctor(cx.tcx, &r).map(|i| i.into_inner())
+        }).collect();
+        let ty = pat_interval.ty;
+        let (pat_interval_lo, pat_interval_hi) = pat_interval.into_inner();
+        for (subrange_lo, subrange_hi) in ranges {
+            if pat_interval_lo > subrange_hi || subrange_lo > pat_interval_hi  {
+                // The pattern doesn't intersect with the subrange at all,
+                // so the subrange remains untouched.
+                remaining_ranges.push(subrange_lo..=subrange_hi);
+            } else {
+                if pat_interval_lo > subrange_lo {
+                    // The pattern intersects an upper section of the
+                    // subrange, so a lower section will remain.
+                    remaining_ranges.push(subrange_lo..=(pat_interval_lo - 1));
+                }
+                if pat_interval_hi < subrange_hi {
+                    // The pattern intersects a lower section of the
+                    // subrange, so an upper section will remain.
+                    remaining_ranges.push((pat_interval_hi + 1)..=subrange_hi);
```

Comment:
You can introduce a closure to push one range, that does all the work below, to avoid having an intermediary `Vec` for `remaining_ranges` - or turn this `for` loop into a `flat_map`, using two chained `option::IntoIter`'s.